substrate/frame/staking/src/slashing.rs

```diff
@@ -341,31 +333,24 @@ fn add_offending_validator<T: Config>(stash: &T::AccountId, disable: bool) {
 
 		let validator_index_u32 = validator_index as u32;
 
-		match offending.binary_search_by_key(&validator_index_u32, |(index, _)| *index) {
-			// this is a new offending validator
-			Err(index) => {
-				offending.insert(index, (validator_index_u32, disable));
-
-				let offending_threshold =
-					T::OffendingValidatorsThreshold::get() * validators.len() as u32;
-
-				if offending.len() >= offending_threshold as usize {
-					// force a new era, to select a new validator set
-					<Pallet<T>>::ensure_new_era()
-				}
-
-				if disable {
-					T::SessionInterface::disable_validator(validator_index_u32);
-				}
-			},
-			Ok(index) => {
-				if disable && !offending[index].1 {
-					// the validator had previously offended without being disabled,
-					// let's make sure we disable it now
-					offending[index].1 = true;
-					T::SessionInterface::disable_validator(validator_index_u32);
-				}
+		let DisablingDecision { disable_offenders } = T::DisablingStrategy::make_disabling_decision(
+			DisablingDecisionContext {
+				offender_idx: validator_index as u32,
+				slash_era: params.slash_era,
+				era_now: params.now,
 			},
+			&disabled,
+			&validators,
+		);
+
+		for offender in disable_offenders {
+			// Add the validator to `DisabledValidators` and disable it. Do nothing if it is
+			// already disabled.
+			if let Err(index) = disabled.binary_search_by_key(&validator_index_u32, |index| *index)
+			{
+				disabled.insert(index, offender);
+				T::SessionInterface::disable_validator(offender);
+			}
 		}
```

Comment:
Can we add a `debug_assert!()` check here to ensure that the result disabled vec is sorted? Something along the lines of:

```rust
debug_assert!(DisabledValidators::<T>::get().is_sorted());
``` 

And also add `try-state` check to ensure this vec is always sorted. I reckon that if at some point for some reason this vec is not sorted, it may lead to a bad security vuln.